File: [H]array/[H]array-two-pointers-opposite-direction/977-squares-of-a-sorted-array.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        res = []
        left, right = 0, len(nums) - 1
        while left <= right:
            if nums[left] ** 2 < nums[right] ** 2:
                res.append(nums[right] ** 2)
                right -= 1
            else:
                res.append(nums[left] ** 2)
                left += 1
        return res[:: - 1]
        
# time O(n)
# space O(n)
# using array and two pointers opposite direction and shrink type

File: [H]array/[H]array-two-pointers-opposite-direction/977-squares-of-a-sorted-array.java
class Solution {
    public int[] sortedSquares(int[] nums) {
        List<Integer> list = new ArrayList<>();
        int left = 0;
        int right = nums.length - 1;
        while (left <= right) {
            if (Math.abs(nums[left]) > Math.abs(nums[right])) {
                list.add(nums[left] * nums[left]);
                left += 1;
            } else {
                list.add(nums[right] * nums[right]);
                right -= 1;
            }
        }
        Collections.reverse(list);
        return list.stream().mapToInt(i -> i).toArray();
    }
}

// time O(n)
// space O(n)
// using array and two pointers opposite direction and shrink type